### PR TITLE
POC: Button component and styling approach

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR is a POC for an approach to implementing and applying styling to `Button` components, with an eye to a sharable, modular set of components for use across apps.

I’m looking for sanity-check level feedback on the high-level component and styling approach.

To review draft new Button components in-browser, check out this branch, run `make dev` and open http://localhost:3000/ui-playground/buttons in your browser. (Don’t take the current state of this UI Playground page as fully-baked; it’s just a working area).

### Input solicited

I’m looking for feedback on the general approach in these two places only:

* `src/sidebar/components/Buttons/index.js` - Button preact components
* `src/styles/sidebar/components/Buttons.scss` - Their SASS styling

### Notes about the state of this code

* the `react/prop-types` ESLint rule is turned off in this branch; placeholder until we do this correctly. The `propTypes` for these `Button` components were intolerable to maintain.
* The pattern library styling and component details should be generally ignored completely here (i.e. don’t review them). I’m using that area truly as a play- and test-ground right now.
* I’ve tossed Rob’s `jsxToString` snippet into this branch because it’s useful for the present moment

### Notes about the structure of the Buttons components module

* The approach used to create the base `ButtonBase` component and the other Button components. Specifically, the use of props spreading and forwarding. We have historically not done this, but it sure allows for some more streamlined code here.
* The `Buttons` module naming is to prevent namespace conflicts for now with `Button`.
* I haven’t yet determined how the `Button`  components will be split into modules (or if they will be).
* `ButtonBase` is not for direct use (obviously, as it’s not exported)
* Each Button component is rendered with _one_ block CSS `class` . Except in the case of `CustomButton`, which takes a `className` prop, that `className` is applied by the component itself.

Put in another way, it’s not possible override the `className` for a `Button` component. “Overriding” is really a “new type of button” and should be handled via the `CustomButton` component.

This may seem overly restrictive, but I believe it could help us keep out of trouble.
* Each `Button` component has the same API surface, and the pattern here is that most props are forwarded on to `ButtonBase`. This is different than what we’ve (ever) done before but rest-props seems to make a lot of sense here and DRYs the hell out of this stuff.

### Notes about the structure of the SCSS

In the same way that nesting/extending is limited in the component code, it is also here: I’ve avoided nesting anything more than one level deep. Bit-by-bit style building in multiple layers of mixins and overrides has proven problematic in our current styling implementation.

I’d like to split this single SASS module into multiple (potentially), but have not yet done that splitting-up. Instead, I’ve commented several “sections” of the module.

One thing that I’d like to call out is the “hoisting” of references to variables from the client’s `variables` SASS module to the very top, and the mapping of those values to some local “tokens”. How we handle these values when this work gets moved into the shared package can be abstracted for now, but the rest of the source does not have a dependency on anything in the client.

To this end, I’ve elected to unroll a few of the simpler util mixins versus introducing dependencies (e.g. flexbox layout, a border on one button). I’m also using `em` units directly here instead of abstracting them through a variable. Striking a balance here so we don’t end up with dependency spaghetti.

The `Button` components’ styling is independent with a caveat: sizing is based on `em`s (as it is, primarily, in the client), which means that Buttons inherit the containing `font-size`. To make Buttons scale to what they’d look like in the client,  I’ve applied a `13px` `font-size` rule to a wrapping element in the UI playground (that’s the client’s `body` font size for better or worse).

### CODE REVIEW

File: src/styles/sidebar/components/Buttons.scss [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    font-size: map.get($base-sizes, 'small');
+  }
+
+  &--size-large {
+    font-size: map.get($base-sizes, 'large');
+  }
+
+  // (icon) SVG sizing is relevant
+  svg {
+    width: 1.25em;
```
Comment (lyzadanger) [inline src/styles/sidebar/components/Buttons.scss:175]: Currently, we use `em` sizing for buttons, but `px` sizing for the icons within them, which is going to lead to some curious results in different combinations. The "average" button in the client has `13px` font (i.e. `1em = 13px`). This is in practice a `1:1.25` ratio `font:icon`.

File: src/styles/sidebar/components/Buttons.scss [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  }
+
+  &--icon-right svg {
+    margin-left: $margin;
+  }
+}
+
+// COMPONENT CLASSES
+
+.IconButton {
```
Comment (lyzadanger) [inline src/styles/sidebar/components/Buttons.scss:209]: Exactly one selector per class here. I think this is important.

**Comment by LMS007** [comment]
Okay, I think this mostly solve much of the issues with the existing button maddess. I think the basic props are good and allow a lot of flexibility. I also like how the aria-* props are much more decoupled from the button styles/props. Also, having the button sandbox is very useful to see the live examples. I'm good with the sass variable mappings which helps isolate the integrations on variables.sass just to the top of the file. 

I'm still trying to get used to the medley of component names with small differences--which is where I'm mostly stuck. I have a strong urge to want to have fewer named components and more prop variances, but I totally get that was the exact opposite goal of this change.  Here are my observations in no particular order:

1. 
"compact" to me feel more like a prop than an entirely different component, but I supposed I could be persuaded.

For example...

`<CompactIconButton>` vs `<IconButton isCompact="true">`  
`< CompactLabeledIconButton >` vs `< LabeledIconButton isCompact="true">`  

2.
LabeledButton  and LabeledIconButton's primary difference is the icon prop, and I just feel like it could be one component with an optional prop, but I get that is not the objective here.  I guess I'm strongly tied to a buffet style thanksgiving plate component rather than al-la-cart / tapas. I do get that if we have a thing called "Labeled" and another thing called "Icon" then a labeled thing should not have an icon thing unless its a third hybrid, e.g. "LabeledIconButton" so I do agree we need the third thing to honor the first 2 things distinctions. I guess I don't have any better suggestions :\

3. 
The component names I see immediate value in are

IconButton
LabeledButton
LinkButton
CustomButton

The IconInputButton is a tricky one. It almost feels strongly coupled to an input field and can't stand up on its own, or at least it feels like it has no use case outside of an input tag. I wonder if we could use a parent component wrapper to bind an icon button and input together and it just re-styles the background color and border? Otherwise, I sorta feel the component should also have an input field, but maybe I'm overthinking it, but to me it does not feel like core button type. It might be okay just as a custom button with just a css class considering it low usage, thoughts?

4.
The `variant` prop seems to only work on the the  LabeledButton, LabeledIconButton and CompactLabeledIconButton. Is there any reason we can't style the background color dark grey and text white for all buttons that use variant="primary" ?  
Otherwise it seems to have a no-op effect in all other cases. 

5.
I'm not seeing the keyboard focus border when tabbing around and I can't figure out how. The code looks correct. I'll have to look closer to figure out why.

**Comment by lyzadanger** [comment]
> LabeledButton and LabeledIconButton's primary difference is the icon prop, and I just feel like it could be one component with an optional prop, but I get that is not the objective here. 

I believe that this kind of merging and melding is part of how we got ourselves in trouble with the existing `Button` component. `LabeledIconButton` is a kind of `LabeledButton` (sort of like an internal "extension") except when it's not and bits and pieces of it have to be overridden. Oh, and, whoops, then we realize that `LabeledIconButton` is actually a layout involving multiple elements (`button`, `svg`) that have to be managed, and, wait, now we need a way to override `LabeledButton` to put an icon on the right, and now we're three mixins deep overriding little pieces of the styles.

The refactor aims to flatten out these kinds of conditional pitfalls. Another aim is for expression of component intent. Consider:

```
<Button buttonText="Edit" title="Do this thing" icon="edit" className="my_slightly_overridden_button_styles" />
```

Quick! What kind of button is that? It might be a labeled-style button with an icon on the left, or not, depending on what liberties that custom CSS class takes. But:

```
<LabeledIconButton icon="edit" title="Do this thing">
  Edit
</LabeledIconButton>
```

or

```
<LabeledButton>
  Edit
</LabeledButton>
```

Allows the developer to see quickly what kind of role the given button is playing.

Differentiating `LabeledIconButton` from `LabeledButton` also allows us to enforce the presence of the `icon` property on the former.

Adding another `prop` to every button increases the complexity across all buttons, while adding another `*Button` component is a templated few-liner. Though I will concede that the balance of what's-a-variant vs. what's-a-component is not black-and-white and I'm still adjusting it.

**Comment by lyzadanger** [comment]
> The variant prop seems to only work on the the LabeledButton, LabeledIconButton and CompactLabeledIconButton. Is there any reason we can't style the background color dark grey and text white for all buttons that use variant="primary" ?
Otherwise it seems to have a no-op effect in all other cases.

A no-op, yep, but a completely benign no-op. We can implement styling when we need it for those components+variant. In fact, in further experimenting on this branch, I've discovered a need for a variant on the `IconButton`.

**Comment by lyzadanger** [comment]
> The IconInputButton is a tricky one. It almost feels strongly coupled to an input field and can't stand up on its own, or at least it feels like it has no use case outside of an input tag. I wonder if we could use a parent component wrapper to bind an icon button and input together and it just re-styles the background color and border?

I'd say both yes and no. The `IconInputButton` might not belong where it is. We may wish to move it. However:

> just re-styles the background color and border

Is antithetical to this approach. I'm pushing back, hard, on nested partial style overrides.

However, I agree with your general principle, that the input+button should be perhaps treated as a wrapped component consisting of both elements. I'd suggest viewing the current `IconInputButton` as somewhat "sketchy" in form and agree that it's not yet baked. I'll take a deeper look at this when I get a chance.

**Comment by lyzadanger** [comment]
> I'm still trying to get used to the medley of component names with small differences--which is where I'm mostly stuck. I have a strong urge to want to have fewer named components and more prop variances,

Can you elaborate on this?

**Comment by LMS007** [comment]
> > LabeledButton and LabeledIconButton's primary difference is the icon prop, and I just feel like it could be one component with an optional prop, but I get that is not the objective here.
> 
> I believe that this kind of merging and melding is part of how we got ourselves in trouble with the existing `Button` component. `LabeledIconButton` is a kind of `LabeledButton` (sort of like an internal "extension") except when it's not and bits and pieces of it have to be overridden. Oh, and, whoops, then we realize that `LabeledIconButton` is actually a layout involving multiple elements (`button`, `svg`) that have to be managed, and, wait, now we need a way to override `LabeledButton` to put an icon on the right, and now we're three mixins deep overriding little pieces of the styles.
> 
> The refactor aims to flatten out these kinds of conditional pitfalls. Another aim is for expression of component intent. Consider:
> 
> ```
> <Button buttonText="Edit" title="Do this thing" icon="edit" className="my_slightly_overridden_button_styles" />
> ```
> 
> Quick! What kind of button is that? It might be a labeled-style button with an icon on the left, or not, depending on what liberties that custom CSS class takes. But:
> 
> ```
> <LabeledIconButton icon="edit" title="Do this thing">
>   Edit
> </LabeledIconButton>
> ```
> 
> or
> 
> ```
> <LabeledButton>
>   Edit
> </LabeledButton>
> ```
> 
> Allows the developer to see quickly what kind of role the given button is playing.
> 
> Differentiating `LabeledIconButton` from `LabeledButton` also allows us to enforce the presence of the `icon` property on the former.
> 
> Adding another `prop` to every button increases the complexity across all buttons, while adding another `*Button` component is a templated few-liner. Though I will concede that the balance of what's-a-variant vs. what's-a-component is not black-and-white and I'm still adjusting it.


Okay, I think that might be a reasonable distinction to warrant a second component name and I agree prop bombing all the buttons with a icon is not ideal and gets us back into the original problem.

**Comment by LMS007** [comment]
> > The variant prop seems to only work on the the LabeledButton, LabeledIconButton and CompactLabeledIconButton. Is there any reason we can't style the background color dark grey and text white for all buttons that use variant="primary" ?
> > Otherwise it seems to have a no-op effect in all other cases.
> 
> A no-op, yep, but a completely benign no-op. We can implement styling when we need it for those components+variant. In fact, in further experimenting on this branch, I've discovered a need for a variant on the `IconButton`.

Fair enough! perhaps we can remove it from the pattern lib at the outset? I don't think there is much value is expressing props that do nothing, I was a bit confused by that because it felt more like a sass mistake. Unless you feel there is good reason to demonstrate the no-op?

**Comment by LMS007** [comment]
> s antithetical to this approach. I'm pushing back, hard, on nested partial style overrides.
> 
> However, I agree with your general principle, that the input+button should b

I agree with the whole nested overrides, it creates obscure dependencies, I think my only concern was this felt very specialized and perhaps could remain a custom button if only used in 2 places?  I could go either way.

**Comment by LMS007** [comment]
> > I'm still trying to get used to the medley of component names with small differences--which is where I'm mostly stuck. I have a strong urge to want to have fewer named components and more prop variances,
> 
> Can you elaborate on this?

Oh, I was just expressing my tendency to keeping everything in one component with a bunch of props because that's what I'm used to... I'm not saying this the best approach at all. Change can be good!

**Comment by esanzgar** [comment]
Other UI component libraries have one or two buttons types and all other subtypes are enabled by props.

https://material.angular.io/components/button/overview
https://material-ui.com/components/buttons/
https://material.preactjs.com/component/button/

I think that makes it simpler to use and remember.

File: src/sidebar/components/Buttons/index.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
+import classnames from 'classnames';
+
+import { SvgIcon } from '@hypothesis/frontend-shared';
+
+/**
+ * @typedef ButtonProps
+ * @prop {Object} [children]
```
Comment (esanzgar) [inline src/sidebar/components/Buttons/index.js:7]:
```suggestion
 * @prop {import("preact").ComponentChildren} [children]
```

File: src/sidebar/components/Buttons/index.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+ * @prop {string} [icon] - name of `SvgIcon` to render in the button
+ * @prop {'left'|'right'} [iconPosition] - Icon positioned to left or to
+ *   right of button text
+ * @prop {boolean} [isDisabled]
```
Comment (esanzgar) [inline src/sidebar/components/Buttons/index.js:11]: Maybe simply `disabled` to mimic the default `<button>` element?

File: src/sidebar/components/Buttons/index.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+/**
+ * @typedef IconButtonBaseProps
+ * @prop {string} icon - Icon is required for icon buttons
+ * @prop {string} title - Title is required for icon buttons
+ */
+
+/**
+ * @typedef {ButtonProps & { className: string }} ButtonBaseProps
+ * @typedef {ButtonProps & IconButtonBaseProps} IconButtonProps
```
Comment (esanzgar) [inline src/sidebar/components/Buttons/index.js:30]: Or like line above:

```suggestion
 * @typedef {ButtonProps & {icon:string, title:string}} IconButtonProps
```

File: src/sidebar/components/Buttons/index.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+/**
+ * A labeled button with no icon
+ * @param {ButtonProps} props
+ */
+export function LabeledButton(props) {
+  const { children } = props;
+  return (
+    <ButtonBase className="LabeledButton" {...props}>
```
Comment (robertknight) [inline src/sidebar/components/Buttons/index.js:114]: This ends up passing `children` to `ButtonBase` in two separate ways. What I suggest doing is `const { children, ...otherProps } = props` and then `<ButtonBase className="LabeledButton" {...otherProps}>`

File: src/sidebar/components/Buttons/index.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+ * @prop {Object} [children]
+ * @prop {string} [icon] - name of `SvgIcon` to render in the button
+ * @prop {'left'|'right'} [iconPosition] - Icon positioned to left or to
+ *   right of button text
+ * @prop {boolean} [isDisabled]
+ * @prop {boolean} [isExpanded] - Is the element associated with this button
+ *   expanded (set `aria-expanded`)
+ * @prop {boolean} [isPressed] - Is this button currently "active?" (set
+ *   `aria-pressed`)
+ * @prop {() => any} [onClick]
```
Comment (esanzgar) [inline src/sidebar/components/Buttons/index.js:16]:
```suggestion
 * @prop {VoidFunction} [onClick]
```

File: src/sidebar/components/Buttons/index.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+ * @prop {'small'|'medium'|'large'} [size] - For styling, size variant
+ * @prop {string} [title]
```
Comment (esanzgar) [inline src/sidebar/components/Buttons/index.js:18]: I think this is clearer
```suggestion
 * @prop {string} [ariaLabel]
```

File: src/sidebar/components/Buttons/index.js [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+import { SvgIcon } from '@hypothesis/frontend-shared';
+
+/**
+ * @typedef ButtonProps
+ * @prop {Object} [children]
+ * @prop {string} [icon] - name of `SvgIcon` to render in the button
+ * @prop {'left'|'right'} [iconPosition] - Icon positioned to left or to
+ *   right of button text
+ * @prop {boolean} [isDisabled]
+ * @prop {boolean} [isExpanded] - Is the element associated with this button
```
Comment (esanzgar) [inline src/sidebar/components/Buttons/index.js:12]:
```suggestion
 * @prop {string} [ariaExpanded] - Is the element associated with this button
```

File: src/sidebar/components/Buttons/index.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+/**
+ * @typedef {ButtonProps & { className: string }} ButtonBaseProps
+ * @typedef {ButtonProps & IconButtonBaseProps} IconButtonProps
+ */
+
+/**
+ * @param {ButtonBaseProps} props
+ */
+function ButtonBase({
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/components/Buttons/index.js:36]: This is meant as a non-usable base component, purposely. The `CustomButton` is what should get used externally for maximum control. `ButtonBase` should not be used directly.

**Comment by lyzadanger** [comment]
I'm working on a new approach based on feedback and thinking...I have some good prototype code and will share it early next week.

**Pull request closed without merging**